improves secure_password in salt/utils/pycrypto.py (#59486)

root360-AndreasUlm opened the pull request:
### What does this PR do?

This PR changes the default character set used by `utils.pycrypto.secure_password()` and implements arguments to control the used character set. (see #59486)

### What issues does this PR fix or reference?
Fixes: None
Closes: #59486
Updated PR for https://github.com/saltstack/salt/pull/59533

### Previous Behavior

`utils.pycrypto.secure_password()` generated random string based on `string.ascii_letters + string.digits`.

### New Behavior

`utils.pycrypto.secure_password()` generates random string based on given character controlling arguments.
The default is now `string.ascii_lowercase + string.ascii_uppercase + string.digits + string.punctuation`.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [X] Docs
- [X] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

root360-AndreasUlm commented:
FYI: This PR was created with pre-commit-hooks disabled due to limitations on development system and hooks not running with containers (yet) (see https://github.com/saltstack/salt/issues/59486#issuecomment-781790748)

s0undt3ch commented on salt/modules/mod_random.py:
```diff
@@ -96,7 +96,7 @@ def str_encode(value, encoder="base64"):
 
 
-def get_str(length=20):
+def get_str(length=20, **kwargs):
```
Please be explicit on the keyword arguments and their defaults. We don't really need this to be `**kwargs`

s0undt3ch commented on salt/modules/mod_random.py:
```diff
@@ -105,13 +105,63 @@ def get_str(length=20):
+        salt '*' random.get_str 128 lowercase=False whitespace=True
     """
-    return salt.utils.pycrypto.secure_password(length)
+    return salt.utils.pycrypto.secure_password(length, **kwargs)
```
as above, pass explicit keyword arguments and their values

s0undt3ch commented on salt/utils/pycrypto.py:
```diff
@@ -41,10 +41,26 @@
 
 
-def secure_password(length=20, use_random=True):
+def secure_password(length=20, use_random=True, **kwargs):
```
Please add the supported keyword arguments and their default values, we don't need this to be `**kwargs`

root360-AndreasUlm commented:
* force rebased the PR on master
* solved linting issues 
* solved coding comments of @s0undt3ch

s0undt3ch committed:
Update docstring

s0undt3ch commented:
I pushed a commit with changes to the docstring

root360-AndreasUlm commented:
> I pushed a commit with changes to the docstring

nice thank you.
I haven't found a documentation about the possible keys to indicate version, yet.
Is there a documentation?

s0undt3ch commented:
The version numbers? the `versionadded`, `versionchanged`?

root360-AndreasUlm commented:
> The version numbers? the `versionadded`, `versionchanged`?

yes those.

s0undt3ch commented:
https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded

root360-AndreasUlm commented:
@s0undt3ch Do all checks have to be successful and the branch always up-to-date for a valid merge request?
I'm currently facing CI-jobs being canceled and multiple times a day I get the message that the branch is out-dated.

s0undt3ch commented:
> Do all checks have to be successful

Yes

> and the branch always up-to-date for a valid merge request?

Not necessarily

> I'm currently facing CI-jobs being canceled and multiple times a day I get the message that the branch is out-dated.

Yes, our Jenkins instance had problems yesterday and there's some instability in our CI jobs which we're trying to address.

root360-AndreasUlm commented:
@s0undt3ch I fixed file conflicts and now all checks have passed.

s0undt3ch committed:
Improve ``secure_password`` in ``salt/utils/pycrypto.py``

Closes #59486

This commit implements arguments for ``utils.pycrypto.secure_password()``
and ``modules.mod_random.get_str()`` to control the character set used by
the string generator.
The arguments kind-of leverage the character sets predefined by
string-class of python.

Signed-off-by: Andreas Ulm <andreas.ulm@root360.de>

s0undt3ch commented:
Conflicts fixed. updated `versionadded`, `versionchanged` to `3004.0`

dwoz reviewed (approved):
I'm good with these changes but `tests/unit/utils/test_pycrypto.py` should be removed if we're not using it.

garethgreenaway committed:
Merge branch 'master' into 59486-defining-chars-get_str
